speed up montecarlo by multiprocessing

Ki41foo opened the pull request.

dickreuter commented:
Great work, I’ll need some time to check it. Have you tried the tests and
made sure they pass? Thanks

On Thu, 28 May 2020 at 21:08, Ki41foo <notifications@github.com> wrote:

> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/dickreuter/Poker/pull/96
> Commit Summary
>
>    - speed up montecarlo by multiprocessing
>
> File Changes
>
>    - *M* poker/decisionmaker/montecarlo_python.py
>    <https://github.com/dickreuter/Poker/pull/96/files#diff-ce08d2e55f192b3112841bb1f65dbd04>
>    (155)
>
> Patch Links:
>
>    - https://github.com/dickreuter/Poker/pull/96.patch
>    - https://github.com/dickreuter/Poker/pull/96.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dickreuter/Poker/pull/96>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJSW7WV4UZP3R5E4G4KLWLRT3AENANCNFSM4NNMNRWQ>
> .
>

Ki41foo commented:
Tested on python 3.8. It should be fine as not much changes.

But note that I removed `passes` and `runs `as condition for timeout since I'm not sure how to break the whole loop in a child process. The workaound is returning `None `and then check if the whole return result list contains `None`, if so raise timeout. A bit dirty but do the trick. Please check if you have better solution

dickreuter commented:
does it work together with the full bot? I've just updated the tables and added a table learner gui (currently not yet open source, but available for free in the binary version).

dickreuter commented:
I think it's a great initiative to make it faster, but it doesn't seem to work with the bot at the moment. I think it will be hard to make changes without me adding the appropriate tests so you can check if it integrates properly.

dickreuter commented:
In the meantime, you may want to have a look at the latest binaries that have been released, you can now teach it new tables: https://sourceforge.net/projects/partypoker-pokerstars-pokerbot/

Ki41foo commented:
what error you're getting?  From my tests it works fine on full bot and its usage is identical as the old one.

dickreuter commented:
Got  a division by zero error. Can check later where exactly.

On Sun, 31 May 2020 at 16:35, Ki41foo <notifications@github.com> wrote:

> what error you're getting? From my tests it works fine on full bot and its
> usage is identical as the old one.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/dickreuter/Poker/pull/96#issuecomment-636487422>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJSW7T7SP3KZZ373QVJ5I3RUJ2NRANCNFSM4NNMNRWQ>
> .
>

Ki41foo commented:
Could be `runs `is 0 if `time.time() > timeout` at the first loop, which's seems impossible.

Anyway, please provide the error details and I'll try to replicate it.

dickreuter commented:
WARNING - Cutting short montecarlo due to timeout
WARNING - Runs: 0
Exception in thread Thread-1:
Traceback (most recent call last):
  File "C:\Users\dickr\Anaconda3\lib\threading.py", line 917, in _bootstrap_inner
    self.run()
  File "C:/Users/dickr/Dropbox/PythonProjects/Poker/poker/main.py", line 178, in run
    m = run_montecarlo_wrapper(p, self.gui_signals, config, ui, t, self.game_logger, preflop_state, h)
  File "C:\Users\dickr\Dropbox\PythonProjects\Poker\poker\decisionmaker\montecarlo_python.py", line 600, in run_montecarlo_wrapper
    ghost_cards=ghost_cards, timeout=timeout, opponent_range=opponent_range)
  File "C:\Users\dickr\Dropbox\PythonProjects\Poker\poker\decisionmaker\montecarlo_python.py", line 469, in run_montecarlo
    self.equity = np.round(wins / runs, 3)
ZeroDivisionError: division by zero

Ki41foo commented:
Can't replicate the error with python3.7

Does it occur even run this file alone? not with full bot.
If not try log `timeout `and `time.time()` in line 448 to check if already timeout at the beginning.

dickreuter commented:
Did you try to run the full bot? It’s possible that multi processing is
actually slower in real life scenario as it has to spawn the processes
first. With 5000 runs that may end up being slower.

On Tue, 2 Jun 2020 at 07:03, Ki41foo <notifications@github.com> wrote:

> Can't replicate the error with python3.7
>
> Does it occur even run this file alone? not with full bot.
> If not try log timeout and time.time() in line 448 to check if already
> timeout at the beginning.
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/dickreuter/Poker/pull/96#issuecomment-637298358>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJSW7XNESUDZYJ6ALKJ2RLRUSI2HANCNFSM4NNMNRWQ>
> .
>

Ki41foo commented:
The bot has been playing 2000+ hands since this change. Never got a timeout or error.

If the main function in this file can work then it should work with full bot too as long as the parameters are resonable. Have you tried run this file soly?

dickreuter commented:
So maybe the parameters are not reasonable. Please rebase and try if you
can get it.to work.

On Tue, 2 Jun 2020 at 08:00, Ki41foo <notifications@github.com> wrote:

> The bot has been playing 2000+ hands since this change. Never got a
> timeout or error.
>
> If the main function in this file can work then it should work with full
> bot too as long as the parameters are resonable. Have you tried run this
> file soly?
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/dickreuter/Poker/pull/96#issuecomment-637332600>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJSW7W2FRBXWKUNXPCINGDRUSPPVANCNFSM4NNMNRWQ>
> .
>

Ki41foo commented:
![Annotation 2020-06-02 150646](https://user-images.githubusercontent.com/8576393/83490580-da48f500-a4e2-11ea-8bee-11b7126fde32.png)

This is the full bot's log. The avg montecarlo runtime is 3~5s

Ki41foo commented:
> So maybe the parameters are not reasonable. Please rebase and try if you can get it.to work.
> […](#)
> On Tue, 2 Jun 2020 at 08:00, Ki41foo ***@***.***> wrote: The bot has been playing 2000+ hands since this change. Never got a timeout or error. If the main function in this file can work then it should work with full bot too as long as the parameters are resonable. Have you tried run this file soly? — You are receiving this because you modified the open/close state. Reply to this email directly, view it on GitHub <[#96 (comment)](https://github.com/dickreuter/Poker/pull/96#issuecomment-637332600)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAJSW7W2FRBXWKUNXPCINGDRUSPPVANCNFSM4NNMNRWQ> .

The parameters I mean `Simulation.run_montecarlo(logging, my_cards, cards_on_table, player_amount=players, ui=None, maxRuns=maxruns,
                              ghost_cards=ghost_cards, timeout=timeout, opponent_range=0.25)`

I have no idea how to debug it since can't replicate the error.

Maybe with rewriting the code in a safer approach later.

dickreuter commented:
Will let you know once I've cleaned up the repo and added the appropriate tests, then it should be easier to merge it and ensure it works together with the rest. I think multiprocessing is a good idea.

Ki41foo commented:
Thanks. I'm looking for a safer way to do this. Will update you once I have
result

On Mon, Jun 8, 2020 at 11:39 PM Nicolas Dickreuter <notifications@github.com>
wrote:

> Will let you know once I've cleaned up the repo and added the appropriate
> tests, then it should be easier to merge it and ensure it works together
> with the rest. I think multiprocessing is a good idea.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/dickreuter/Poker/pull/96#issuecomment-640707858>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACBN3CMHCIVU2BIYXSA73DTRVUA2RANCNFSM4NNMNRWQ>
> .
>

Ki41foo commented:
The issue may be related to computer OS as processes are much heavy in
windows, so the overhead exceeds timeout.
Maybe we can init the pool at first once for all and don't close it since
our main program is a loop

On Tue, Jun 9, 2020 at 12:55 AM Kai Li <ki41foo@gmail.com> wrote:

> Thanks. I'm looking for a safer way to do this. Will update you once I
> have result
>
> On Mon, Jun 8, 2020 at 11:39 PM Nicolas Dickreuter <
> notifications@github.com> wrote:
>
>> Will let you know once I've cleaned up the repo and added the appropriate
>> tests, then it should be easier to merge it and ensure it works together
>> with the rest. I think multiprocessing is a good idea.
>>
>> —
>> You are receiving this because you modified the open/close state.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/dickreuter/Poker/pull/96#issuecomment-640707858>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACBN3CMHCIVU2BIYXSA73DTRVUA2RANCNFSM4NNMNRWQ>
>> .
>>
>
